spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java

```diff
@@ -39,14 +112,17 @@
 		if (payload instanceof Throwable) {
 			Throwable cause = (Throwable) payload;
 			while (cause != null) {
-				String causeName = cause.getClass().getName();
-				if (this.getChannelMappings().keySet().contains(causeName)) {
-					mostSpecificCause = causeName;
+				for (Map.Entry<String, Class<?>> entry : this.classNameMappings.entrySet()) {
+					String channelKey = entry.getKey();
+					Class<?> exceptionClass = entry.getValue();
+					if (exceptionClass.isInstance(cause)) {
+						mostSpecificCause = channelKey;
+					}
 				}
 				cause = cause.getCause();
 			}
 		}
-		return Collections.singletonList((Object) mostSpecificCause);
+		return mostSpecificCause != null ? Collections.<Object>singletonList(mostSpecificCause) : null;
```

Comment:
This is not really needed; the caller handles null entries. I'm ok to leave it though.